mayastor/tests/common/compose.rs

```diff
@@ -0,0 +1,587 @@
+    },
+    errors::Error,
+    network::ListNetworksOptions,
+    service::{
+        ContainerSummaryInner, EndpointIpamConfig, EndpointSettings,
+        HostConfig, Ipam, Mount, MountTypeEnum, Network,
+    },
+    Docker,
+};
+use futures::TryStreamExt;
+use ipnetwork::Ipv4Network;
+use tokio::sync::oneshot::channel;
+use tonic::transport::Channel;
+
+use ::rpc::mayastor::{
+    bdev_rpc_client::BdevRpcClient, mayastor_client::MayastorClient,
+};
+use mayastor::core::{
+    mayastor_env_stop, MayastorCliArgs, MayastorEnvironment, Reactor, Reactors,
+};
+
+#[derive(Clone)]
+pub struct RpcHandle {
+    pub name: String,
+    pub endpoint: SocketAddr,
+    mayastor: MayastorClient<Channel>,
+    pub bdev: BdevRpcClient<Channel>,
+}
+
+impl RpcHandle {
+    /// connect to the containers and construct a handle
+    async fn connect(name: String, endpoint: SocketAddr) -> Self {
+        loop {
+            if TcpStream::connect_timeout(&endpoint, Duration::from_millis(100))
+                .is_ok()
+            {
+                break;
+            } else {
+                thread::sleep(Duration::from_millis(101));
+            }
+        }
+
+        let mayastor =
+            MayastorClient::connect(format!("http://{}", endpoint.to_string()))
+                .await
+                .unwrap();
+        let bdev =
+            BdevRpcClient::connect(format!("http://{}", endpoint.to_string()))
+                .await
+                .unwrap();
+
+        Self {
+            name,
+            mayastor,
+            bdev,
+            endpoint,
+        }
+    }
+}
+
+pub struct Builder {
+    /// name of the experiment this name will be used as a network and labels
+    /// this way we can "group" all objects within docker to match this test
+    /// test. It is highly recommend you use a sane name for this as it will help you
+    /// during debugging
+    name: String,
+    /// containers we want to create, note these are mayastor containers
+    /// only
+    containers: Vec<ContainerName>,
+    /// the network for the tests used
+    network: String,
+    /// delete the container and network when dropped
+    clean: bool,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Builder::new()
+    }
+}
+
+impl Builder {
+    /// construct a new builder for `[ComposeTest']
+    pub fn new() -> Self {
+        Self {
+            name: "".to_string(),
+            containers: Default::default(),
+            network: "10.1.0.0".to_string(),
+            clean: false,
+        }
+    }
+
+    /// set the network for this test
+    pub fn network(mut self, network: &str) -> Builder {
+        self.network = network.to_owned();
+        self
+    }
+
+    /// the name to be used as labels and network name
+    pub fn name(mut self, name: &str) -> Builder {
+        self.name = name.to_owned();
+        self
+    }
+
+    /// add a mayastor container with a name
+    pub fn add_container(mut self, name: &str) -> Builder {
```

Comment:
would having another call with a config argument be useful? That way we could start the container with the appropriate config.

Reply:
You mean a YAML config right? If so then yes, but not sure yet when as the config work is a bit up in the air.. that is, how we use it. For manual testing the config can be used at all times. Doing explicit rpc is more work OTOH, it does provide better coverage..

Reply:
Sorry, yes the YAML config. Yeah good point, I guess with grpc it follows a more stable "contract".